### ISSUE TITLE
Current slashing conditions are not sufficient

### ISSUE THREAD

**Issue by josojo** [ethereum/eth2.0-specs#362]
I think that I have found a bug, @djrtwo confirmed it on gitter.

The current slashing conditions are fine for normal finalizations without slots between potential justifiable/finalize-able blocks (i.e epoch_boundary blocks) or if shuffling is deterministic within the context of a finalized block.
But, in the current spec, neither of these two conditions are met. We allow attestation between epoch_boundary blocks and a reshuffling is possible.

I tried to describe it more detailed here:
https://docs.google.com/presentation/d/10gODMelMaeZzKgRuBk1G0At-wwC5VcCkvBoMhkHQU80/edit?usp=sharing

----------------------------------------
As a solution, I would propose to change the current is_double_vote to something like this:

```
def is_double_vote(attestation_data_1: AttestationData,
                   attestation_data_2: AttestationData) -> bool
    """
    Assumes ``attestation_data_1`` is distinct from ``attestation_data_2`` 
    Returns True if the provided ``AttestationData`` are slashable
    due to a 'double vote'.
    """
    return attestation_data_1.epoch_boundary_root != attestation_data_2.epoch_boundary_root AND  attestation_data_1.slot // EPOCH_LENGTH == attestation_data_2.slot // EPOCH_LENGTH
```

**Comment by djrtwo** [comment]
The fix should be the following:

```python
def is_double_vote(attestation_data_1: AttestationData,
                   attestation_data_2: AttestationData) -> bool
    """
    Assumes ``attestation_data_1`` is distinct from ``attestation_data_2`` 
    Returns True if the provided ``AttestationData`` are slashable
    due to a 'double vote'.
    """
    return attestation_data_1.slot // EPOCH_LENGTH == attestation_data_2.slot // EPOCH_LENGTH

**Comment by josojo** [comment]
Second thought, let's have the fix like this:

```Python
def is_double_vote(attestation_data_1: AttestationData,
                   attestation_data_2: AttestationData) -> bool
    """
    Assumes ``attestation_data_1`` is distinct from ``attestation_data_2`` 
    Returns True if the provided ``AttestationData`` are slashable
    due to a 'double vote'.
    """
    return height(attestation_data_1.epoch_boundary_root) == height(attestation_data_2.epoch_boundary_root)
```
(This covers the case where validators want to attest to an epoch_boundary_block from an older epoch than the current epoch)

**Comment by djrtwo** [comment]
`slot` is sufficient

We need an attestation to be slashable without having to have data about any forks of the chain. The `height(boundary_root)` function requires a reverse lookup from hash to slot for both of the attestation messages. For at least one of the messages, the boundary root will not be be in current chain (at most one block per slot in any chain).

**Comment by djrtwo** [comment]
Closed via #373
